dimod/core/initialized.py

```diff
@@ -0,0 +1,204 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# =============================================================================
+
+import abc
+
+from collections import namedtuple
+from numbers import Integral
+
+import numpy as np
+
+from dimod.sampleset import as_samples, infer_vartype, SampleSet
+from dimod.vartypes import Vartype
+
+__all__ = ['Initialized']
+
+
+_parsed = namedtuple('ParsedInputs',
+                     ['initial_states',
+                      'initial_states_generator',
+                      'num_reads',
+                      'seed'])
+
+
+class Initialized(abc.ABC):
+
+    # Allows new generators to be registered
+    _generators = {}
+
+    def parse_initial_states(self, bqm,
+                             initial_states=None,
+                             initial_states_generator='random',
+                             num_reads=None, seed=None):
+        """Parses/generates initial states for an initialized sampler.
+
+        Args:
+            bqm (:class:`~dimod.BinaryQuadraticModel`):
+                The binary quadratic model to be sampled.
+
+            num_reads (int, optional, default=len(initial_states) or 1):
+                Number of reads. If `num_reads` is not explicitly given, it is
+                selected to match the number of initial states given.
+                If initial states, it is defaulted to 1.
+
+            initial_states (samples-like, optional, default=None):
+                One or more samples, each defining an initial state for all the
+                problem variables. Initial states are given one per read, but
+                if fewer than `num_reads` initial states are defined,
+                additional values are generated as specified by
+                `initial_states_generator`. See func:`.as_samples` for a
+                description of "samples-like".
+
+            initial_states_generator ({'none', 'tile', 'random'}, optional, default='random'):
+                Defines the expansion of `initial_states` if fewer than
+                `num_reads` are specified:
+
+                * "none":
+                    If the number of initial states specified is smaller than
+                    `num_reads`, raises ValueError.
+
+                * "tile":
+                    Reuses the specified initial states if fewer than
+                    `num_reads` or truncates if greater.
+
+                * "random":
+                    Expands the specified initial states with randomly
+                    generated states if fewer than `num_reads` or truncates if
+                    greater.
```

Comment:
I am of two minds, I did [call generator args by keyword](https://github.com/dwavesystems/dimod/pull/644/files#diff-7305cc893ce1327fee8a6abc8aeae935R137), which combined with your suggestion of `**kwargs` above, would allow us to provide a very general callable. On the other hand I am wondering if we want this to simmer a bit more before generalizing too much.

That's also why `Initialzed._generators` is still "private".

Reply:
For the other suggestion, I like `low`/`high`.

Reply:
Though, I think dwave-hybrid uses [min](https://github.com/dwavesystems/dwave-hybrid/blob/15b2bcaea2e55943b9fcb1c13720776b0a07f0f6/hybrid/utils.py#L516)/[max](https://github.com/dwavesystems/dwave-hybrid/blob/15b2bcaea2e55943b9fcb1c13720776b0a07f0f6/hybrid/utils.py#L538) so maybe to be consistent?

Reply:
Sure, we can let it simmer. I was never quite happy with min/max.